[BEAM-11173] Add Bigtable table with read operation

piotr-szuberski opened the pull request:
Added Bigtable table provider with simple read operation - filters and write will be done in another one in the near future.

Problems:
BeamSQL doesn't support `ARRAY<ROW>`. If you use
```
CREATE TABLE test(
  key VARCHAR,
  family ROW<q1 ARRAY<ROW<val VARCHAR, timestampMicros BIGINT>>>
)
```

and then

`SELECT t.key, t.family.q1  FROM test t`

Calcite shouts that it couldn't parse the generated code.

What is possible is of little value:
`SELECT t.key, t.family.q1[0] FROM test t`

So it's blocked to select all cells with timestamp and labels for now, but as I read about Bigtable it's usually not the desired case (the schema advice is to have single value per cell and usually only the latest value is needed).

I used `val` instead of `value` because value is a keyword in SQL and hence would need to be used with backslashes.

For the Flat approach I followed the Spark's approach [1]

For the Row schema I combined Flink's approach [2] (with only latest values) and Bigquery's approach [3] (with possibility of using ROW)

[1] https://stackoverflow.com/questions/39530938/sparksql-on-hbase-tables
[2] https://ci.apache.org/projects/flink/flink-docs-stable/dev/table/connectors/hbase.html
[3] https://cloud.google.com/bigquery/docs/reference/rest/v2/tables#bigtableoptions

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [![Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.


GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/workflows/Build%20python%20source%20distribution%20and%20wheels/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/workflows/Python%20tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/workflows/Java%20Tests/badge.svg?branch=master&event=schedule)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

piotr-szuberski commented:
I'm having a quite hard time with raw protobuf problem. It's forbidden to use raw protobuf in `sdks:java:extensions:sql` but Bigtable requires it if I want to write tests.

I can't use the pubsub's approach and move Bigtable table provide to `:sdks:java:io:gcp` because later I'll need BeamSqlTableFilter which is in beam sql extensions.

A solution would be to create some testing utility that wraps the BigtableEmulator and bigtable Row creation. But it would have to be public to be able to use it from the SQL module.

amaliujia commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/bigtable/BigtableRowToBeamRow.java:
```diff
@@ -0,0 +1,160 @@
+@SuppressWarnings({
+  "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+})
+public class BigtableRowToBeamRow
```
Nit: it might be better to list the BigTable reference: https://cloud.google.com/bigtable/docs/reference/data/rpc/google.bigtable.v2

piotr-szuberski replied:
Done.

amaliujia commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/bigtable/BigtableRowToBeamRow.java:
```diff
@@ -0,0 +1,160 @@
+@SuppressWarnings({
+  "nullness" // TODO(https://issues.apache.org/jira/browse/BEAM-10402)
+})
+public class BigtableRowToBeamRow
```
Can you add java doc here to give briefly design decision for mapping from bigtable row to Beam row.

I see that bigtable row contains 
Row: Key, Family[]
Family: ... Column[]
Column: ... Cell[]
Cell:  ...Value

document current mapping seems will be straightforward?

piotr-szuberski replied:
I've written some javadoc, but I'm not sure whether it's understable

amaliujia reviewed (commented):
Thanks for working on this.

My current question is, for the problem you find that ARRAY<ROW> is not well supported, what part of design have it affected?

Basically my view is, we should support ARRAY<ROW> rather than make it change the design to represent bigtable data into Beam schema.

piotr-szuberski commented:
> My current question is, for the problem you find that ARRAY is not well supported, what part of design have it affected?

It affects the possibility of reading multiple cells with more fields than `value`, e.g.:
```
key VARCHAR,
family1 ROW<
  column1 ARRAY<ROW<val BOOLEAN, timestampMicros BIGINT>>
>
```
For now it's possible only to get the array of cell's values.

In general Flink and Spark don't support the array of values. Instead they provide only recent values without the metadata like `timestampMicros` or `labels`.

The Flink's approach is already implemented.

Spark's approach is to skip families and columns in the schema definition, provide just values and then add mapping to each value in the TBLPROPERTIES in format 
```
schemaField1 BIGINT,
schemaField2 VARCHAR
TBLPROPERTIES { "hbase.columns.mapping": "familyName:columnName:schemaField1,familyName:columnName:schemaField2"`
```

On the other hand BigQuery does support it.

Maybe for the simplicity's sake should we just stick to the Flink and Spark version?

I can see that in Pubsub table for some reason it's possible to use only flattened schema for the Write operation. If it's not possible to do so with ROW then Spark's approach would fit well for it.

piotr-szuberski commented:
Run SQL PostCommit

piotr-szuberski commented:
@amaliujia I've added some javadoc but I'm not sure about it.

I've also added flat schema (like Spark does [1])so it should be possible to execute write operations in the near future.

[1] https://stackoverflow.com/questions/39530938/sparksql-on-hbase-tables

amaliujia commented:
Thank you! I will take a another look

piotr-szuberski commented:
I've moved the transforms to io/bigtable because Write transforms need to import `com.google.protobuf.ByteString` which is not allowed in extensions/sql

piotr-szuberski commented:
@amaliujia

amaliujia commented on sdks/java/extensions/sql/src/test/java/org/apache/beam/sdk/extensions/sql/meta/provider/bigtable/BigtableTableCreationFailuresTest.java:
```diff
@@ -0,0 +1,130 @@
+  private static String location() {
+    return "googleapis.com/bigtable/projects/fakeProject/instances/fakeInstance/tables/beamTable";
+  }
+}
```
Add a test to hit 
```
  private static void validateMatcher(Matcher matcher, String location) {
    if (!matcher.matches()) {
      throw new InvalidTableException(
          "Bigtable location must be in the following format:"
              + " 'googleapis.com/bigtable/projects/projectId/instances/instanceId/tables/tableId'"
              + " but was: "
              + location);
    }
  }
```
?

piotr-szuberski replied:
Done.

amaliujia commented on website/www/site/content/en/documentation/dsls/sql/extensions/create-external-table.md:
```diff
@@ -204,6 +205,131 @@ TYPE bigquery
+    *   `VARCHAR`
+    *   `BIGINT`
+    *   `DOUBLE`
+    *   `BOOLEAN`
```
So what is the reason that single type only supports these 5 types (i.e. not include other SQL types like TIMESTAMP)?

piotr-szuberski replied:
BigQuery supports only those 5 types. But I agree that there is no reason why not to support others as well. Should I do that in this PR or in the next one?

amaliujia replied:
Maybe just do that in this PR to make this feature complete?

piotr-szuberski replied:
Ok, will do!

piotr-szuberski commented on sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/bigtable/BigtableTable.java:
```diff
@@ -0,0 +1,200 @@
+    return BeamTableStatistics.BOUNDED_UNKNOWN;
+  }
+
+  private static Map<String, List<String>> parseColumnsMapping(String commaSeparatedMapping) {
```
Done.

piotr-szuberski replied:
Oh, I've remembered why they were static. That's because of null checker - it refuses to use non-static functions in the constructor (got required @Initialized and @NonNull BigtableTable)

piotr-szuberski commented:
@amaliujia I've moved the logic of parsing cell value to CellValueParser and added unit tests for parsing other BeamSQL types.

piotr-szuberski commented:
Run Java_Examples_Dataflow PreCommit

piotr-szuberski commented:
@amaliujia

amaliujia commented:
LGTM
